### ISSUE TITLE
Data layout specification is incompatible with atomic loads and stores

### ISSUE THREAD

**Issue by shepmaster** [avr-llvm/llvm#214]
Rust generates atomic loads and stores using the alignment specified by the data layout. However, [it's forbidden](https://github.com/avr-llvm/llvm/blob/2860030806d0fb22831192835e2e257622d73575/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp#L3950-L3951) to load from e.g. a 16-bit value with less than 16-bits of alignment.

This can be fixed by changing the data layout:

```
e-p:16:16:16-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-n8
```

Although I'm not sure what all potential fallout this might have (#213 maybe?)

**Comment by rjordans** [comment]
Do we want/have the floating point specified in the above data layout yet?

**Comment by shepmaster** [comment]
@rjordans ¯\_(ツ)_/¯ It's [already there in some form](https://github.com/avr-llvm/llvm/blob/2860030806d0fb22831192835e2e257622d73575/lib/Target/AVR/AVRTargetMachine.cpp#L47).

**Comment by dylanmckay** [comment]
This seems like a good idea. Would like to wait until you updated #213 so I can check it fixes it.

**Comment by shepmaster** [comment]
> Would like to wait until you updated #213 so I can check it fixes it.

Oh, #213 happens *after* I've made this change locally; I'm worried that this change causes that issue. Still trying to reproduce and reduce that issue :-)

**Comment by Amanieu** [comment]
Since avr doesn't support atomic operations you should add "max-atomic-width": 0 to the Rust target specification. This will disable atomic types in libcore.

**Comment by shepmaster** [comment]
> Since avr doesn't support atomic operations you should add "max-atomic-width": 0 to the Rust target specification. This will disable atomic types in libcore.

We actually have [some minimal emulation of atomic types](https://github.com/avr-llvm/llvm/pull/211): disable interrupts, perform the operation, re-enable interrupts.

**Comment by dylanmckay** [comment]
Fixed in 5f12f20bafa50b178395177cb98aef7de5f68e73
